<a href="https://colab.research.google.com/github/VitalyGladyshev/LLM-engineering/blob/main/LLM_GLVV_HW11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Библиотеки и импорты

In [6]:
!pip install python-docx

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 253.0/253.0 kB 15.5 MB/s eta 0:00:00


In [7]:
!pip install langchain-community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 48.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 46.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.2/471.2 kB 23.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.8 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.79
    Uninstalling langchain-core-0.3.79:
      Successfully uninstalled langchain-core-0.3.79
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 0.3.11
    Uninstalling langchain-text-splitters-0.3.11:
      Successfully uninstalled langchain-text-splitters-0.3.11
ERROR: pip's dependency resolver

In [1]:
!pip install langchain-huggingface

In [2]:
!pip install langchain-chroma

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 5.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.8/20.8 MB 88.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 21.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 71.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.3/103.3 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.4/17.4 MB 90.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.5/72.5 kB 7.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.3/132.3 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.9/65.9 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 208.0/208.0 kB 17.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 105.4/105.4 kB 9.6 MB/s eta 

In [3]:
!pip install pymystem3

In [75]:
!pip install rank_bm25

In [172]:
!pip install ragas datasets nltk rouge-score sentence-transformers

  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of langchain-community to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of langchain-openai to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 337.0/337.0 kB 25.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 160.9/160.9 kB 13.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 50.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 449.8/449.8 kB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 66.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.0/76.0 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 90.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━

In [18]:
import os

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import random
import re

import math
import json
from pathlib import Path

from time import sleep

import docx
from docx import Document
from docx.text.paragraph import Paragraph

from tqdm import tqdm

import time

from pymystem3 import Mystem

random.seed(42)

# Задание 1 Сгенерировать Golden dataset

Составить описание процесса сбора датасета; описать, какую часть вопросов из ретрива данный золотой набор покрывает и почему; объяснить выбор размера золотого набора

In [19]:
data = pd.DataFrame(columns=['header','text', 'question'])
data

Empty DataFrame
Columns: [header, text, question]
Index: []

In [20]:
def save_to_dataframe(file_path):
    global data

    doc = Document(file_path)
    curr_par_len = 0
    chapters = []

    lvl = 0
    number_arr = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    for item in doc.iter_inner_content(): # Use iter_inner_content for robust iteration
        if isinstance(item, Paragraph):
            paragraph = item
            style = paragraph.style.name

            if "Heading" in style:

                hd_parts = style.split()
                curr_lvl = int(hd_parts[-1]) - 1
                number_arr[curr_lvl] += 1

                # Проверим на заполнение номеров предыдущих уровней
                for level in range(curr_lvl):
                    if not number_arr[level]:
                        number_arr[level] = 1

                number = ""
                for i in range(curr_lvl+1):
                    if number:
                        number += "."
                    number += str(number_arr[i])

                while curr_lvl < lvl:
                    number_arr[lvl] = 0
                    lvl -= 1

                lvl = curr_lvl

                print(f"Добавлена глава: {number} {paragraph.text.strip()}")
                curr_par_len = 0

                header = paragraph.text.strip()
                if type(header) == type(0.0):
                    header = ""
                chapters.append([number + " " + header, ""])

            elif paragraph.text:
                curr_par_len += len(paragraph.text)
                if len(chapters):
                    chapters[-1][-1] += paragraph.text
                else:
                    chapters.append([paragraph.text[:20], paragraph.text])

        elif isinstance(item, docx.table.Table):
            # Process table
            table_data = []
            for row in item.rows:
                row_data = [cell.text.strip() for cell in row.cells]
                table_data.append(row_data)

            if len(chapters):
                chapters[-1][-1] += str(table_data)
            else:
                chapters.append([str(table_data)[:20], str(table_data)])

    for chapter in chapters:
        new_row = {'header': chapter[0], 'text': chapter[1]}
        data = pd.concat([data,
                          pd.DataFrame([new_row], columns=data.columns)],
                         ignore_index = True)
    return data

In [23]:
data = save_to_dataframe("/content/merged_cr.docx")

Добавлена глава: 1 Назначение программного модуля CAD
Добавлена глава: 2 Условия выполнения программного модуля CAD
Добавлена глава: 3 Выполнение программного модуля CAD
Добавлена глава: 3.1 Запуск программного модуля CAD
Добавлена глава: 3.2 Управление приложениями
Добавлена глава: 3.3 Работа с документами
Добавлена глава: 3.3.1 Создание нового документа
Добавлена глава: 3.3.2 Открытие документов
Добавлена глава: 3.3.3 Сохранение документов
Добавлена глава: 3.3.4 Закрытие документа
Добавлена глава: 3.3.5 Настройка документов
Добавлена глава: 3.3.5.1 Свойства документа
Добавлена глава: 3.3.5.2 Атрибуты документа
Добавлена глава: 3.3.5.3 История документа
Добавлена глава: 3.3.5.4 Стили объектов
Добавлена глава: 3.4 Справочная система
Добавлена глава: 4 Интерфейс
Добавлена глава: 4.1 Окна программного модуля CAD
Добавлена глава: 4.1.1 Рабочие окна
Добавлена глава: 4.1.2 Вспомогательные окна
Добавлена глава: 4.2 Инструментальные панели
Добавлена глава: 4.2.1 Панель «Лента»
Добавлена глава

In [24]:
data.head()

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2          3 Выполнение программного модуля CAD   
3            3.1 Запуск программного модуля CAD   
4                   3.2 Управление приложениями   

                                                text question  
0  Программный модуль CAD разработан на базе инте...      NaN  
1  Достаточными условиями выполнения программного...      NaN  
2                                                         NaN  
3  Загрузка и запуск программного модуля CAD осущ...      NaN  
4  Данное окно предназначено для осуществления уп...      NaN

In [25]:
data = data.loc[data.text != ""]
data.reset_index(drop=True, inplace=True)

In [4]:
from openai import OpenAI

from google.colab import userdata

MODEL_CONFIG = {
    "base_url": "https://api.vsegpt.ru/v1",
    "api_key": userdata.get('OPENAI_API_KEY'),
    "model_name": "openai/gpt-oss-120b-fast", # "openai/gpt-5", #-mini", # "openai/gpt-5-nano"
    "temperature": 0.3,
    "top_p": 0.9
}

In [32]:
class LLM:
    def __init__(self):
        self.llm = OpenAI(base_url=MODEL_CONFIG["base_url"], api_key=MODEL_CONFIG["api_key"])
        self.system_prompt = "Ты выполняеешь задание пользователя. Не добавляй лишних слов, выводи только результат"

    def set_system_prompt(self, system_prompt: str):
        """Установка системного промпта"""
        self.system_prompt = system_prompt

    def generate_answer(self, question: str) -> str:
        """Генерация ответа модели"""

        # no_think = "/no_think\n"
        messages = []
        messages.append({"role": "system", "content": self.system_prompt})
        messages.append({"role": "user", "content": question}) # no_think +

        reply = self.llm.chat.completions.create(
            model=MODEL_CONFIG["model_name"],
            messages=messages,
            temperature=MODEL_CONFIG["temperature"],
            top_p=MODEL_CONFIG["top_p"],
            stream=False
        )

        return reply.choices[0].message.content #[19:]

In [33]:
llm = LLM()

In [34]:
llm.generate_answer("Напиши последовательно числа от 1 до 10")

'1 2 3 4 5 6 7 8 9 10'

In [39]:
llm.set_system_prompt(
            "Ты участвуешь в решении задачи оценки конвейра RAG. "
            "Ты создаёшь набор данных для оценки семантического поиска. "
            "Задай вопрос к тексту с использованием лексики близкой, но неидентичной лексике текста, "
            "так чтобы протестировать систему поиска - чтобы сравнивать различные алгоритмы поиска. "
            "Выводи только текст вопроса. Не добавляй ничего лишнего!"
        )

In [40]:
for i in tqdm(range(data.shape[0])):
    if data.iloc[i]["text"]:
        prompt = (
            f"Представь себя пользователем RAG системы. "
            f"Создай вопрос к тексту с использованием лексики близкой, но неидентичной лексике текста."
            f"Вопрос должен быть сложным для системы поиска, но выполнимым - чтобы сравнивать различные алгоритмы поиска."
            f"Выводи только текст вопроса. Не добавляй ничего лишнего!"
            f"\nТекст:\n\n{data.iloc[i]['text']}"
        )

        data.loc[i, 'question'] = llm.generate_answer(prompt)

100%|██████████| 40/40 [01:04<00:00,  1.61s/it]


In [41]:
data.head(10)

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   
5               3.3.1 Создание нового документа   
6                     3.3.2 Открытие документов   
7                   3.3.3 Сохранение документов   
8                      3.3.4 Закрытие документа   
9                    3.3.5.1 Свойства документа   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   
5  Команда «Новый документ»  предназначена для со...   
6  Команда «Открыть»  вызывает стандартный диалог...   
7  Команда «Сохранить»  выполняет сохранение теку...   
8  Для закрытия документа нажмите Х на вкладке ря...   
9  Свойства активного документа отображаются во в...   

                                            question  
0  В чём состоит функциональное назначение CAD‑мо...  
1  Какие минимальные программные и аппаратные хар...  
2  Каким способом пользователь может открыть прив...  
3  Каким образом можно вызвать диалог управления ...  
4  В каком пункте меню отображается список средст...  
5  Каким способом можно включить пользовательский...  
6  Каким способом можно запустить открытие RGP‑фа...  
7  Как установить интервал автосохранения и гаран...  
8  Каким образом можно завершить работу с текущим...  
9  Каким способом в окне свойств текущего файла м...

In [42]:
data.to_csv("parsed_qa.csv", index=False)

## Описание

Парсинг текста выполнен по главам - с разбивкой по заголовкам. Таким образом выполнено деление на смысловые части. К каждой главе задаётся вопрос для тестирования.Всего 40 глав и 40 вопросов.

# Задание 2 Протестировать ретриверы на золотом наборе

Объяснить, по какому принципу считали "попадание" в контекст, какие метрики выбрали для оценки качества ретрива

In [44]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

import torch
from langchain_core.documents import Document
from langchain_huggingface.embeddings import HuggingFaceEmbeddings
from langchain_chroma import Chroma
from chromadb import PersistentClient

from chromadb.utils import embedding_functions
from langchain_huggingface.embeddings import HuggingFaceEmbeddings

from sentence_transformers import SentenceTransformer

import chromadb
from uuid import uuid4

In [45]:
SIZE = 300

Протестировать различные ретриверы на данных, обосновать выбор

In [46]:
data_chunks = pd.DataFrame(columns=['id_ch','text_ch', 'text_ch_prep', 'chapter_index', 'part_index'])
data_parts = pd.DataFrame(columns=['id_part','text_part'])

In [130]:
content2index = {}

In [131]:
documents = []
metadatas = []
ids = []

part_cnt = 0

for i in range(data.shape[0]):

    text_len = len(data.iloc[i]["text"])

    if data.iloc[i]["text"] == "-" or data.iloc[i]["text"] != data.iloc[i]["text"]:
        continue

    elif text_len <= SIZE:
        print(f"{i} длина: {len(data.iloc[i]['text'])}")

        part_id = str(uuid4())
        chunk_id = str(uuid4())

        documents.append(f"{data.iloc[i]['header'].strip()}\n{data.iloc[i]['text'].strip()}")
        metadatas.append({"header": data['header'][i],
                          "index": str(i),
                          "chunk":  "",
                          "content": data.iloc[i]["text"].strip(),
                          "id_ch": chunk_id,
                          "chapter_index": str(i)
                         })
        ids.append(chunk_id)
        content2index[data.iloc[i]["text"].strip()] = i
        new_row = {'id_part': part_id, 'text_part': data.iloc[i]["text"].strip()}
        data_parts = pd.concat([data_parts,
                                pd.DataFrame([new_row], columns=data_parts.columns)],
                               ignore_index = True)

        new_row = {'id_ch': str(uuid4()),
                   'text_ch': data.iloc[i]["text"].strip(),
                   'text_ch_prep': "",
                   'part_index': part_id,
                   'chapter_index': str(i)}
        data_chunks = pd.concat([data_chunks,
                                pd.DataFrame([new_row], columns=data_chunks.columns)],
                               ignore_index = True)
        part_cnt += 1

    else:
        print(f"{i}")
        text_splitter = RecursiveCharacterTextSplitter(
            chunk_size=SIZE,
            chunk_overlap=0,
            separators=["\n", " ", ""]  # "\n\n", Порядок приоритета разделителей
        )

        chunks = text_splitter.split_text(data.iloc[i]["text"].strip())

        len_ch = len(chunks)
        step = 13
        if len_ch > step:
            for j in range(step, len_ch, step):
                part = " ".join(chunks[j-(step):j])
                part_id = str(uuid4())
                for k in range(j - step, j):
                    chunk_id = str(uuid4())

                    documents.append(f"{data['header'][i].strip()}\n{chunks[k].strip()}")
                    metadatas.append({"header": data['header'][i],
                                      "index": str(i),
                                      "chunk":  chunks[k],
                                      "content": part.strip(),
                                      "id_ch": chunk_id,
                                      "chapter_index": str(i)
                                     })
                    ids.append(chunk_id)
                    content2index[part.strip()] = i

                    new_row = {'id_ch': str(uuid4()),
                               'text_ch': chunks[k],
                               'text_ch_prep': "",
                               'part_index': part_id,
                               'chapter_index': str(i)}
                    data_chunks = pd.concat([data_chunks,
                                            pd.DataFrame([new_row], columns=data_chunks.columns)],
                                            ignore_index = True)
                    print(f"part len: {len(part)} chunk len: {len(chunks[k])}") # {chunks[k]:50}")

                new_row = {'id_part': part_id, 'text_part': part.strip()}
                data_parts = pd.concat([data_parts,
                                        pd.DataFrame([new_row], columns=data_parts.columns)],
                                    ignore_index = True)
                part_cnt += 1
        else:
            j = 0

        if j < len_ch or len_ch < step:
            part = " ".join(chunks[j:])
            part_id = str(uuid4())
            for k in range(j, len_ch):
                chunk_id = str(uuid4())

                documents.append(f"{data['header'][i].strip()}\n{chunks[k].strip()}")
                metadatas.append({"header": data['header'][i],
                                  "index": str(i),
                                  "chunk":  chunks[k],
                                  "content": part.strip(),
                                  "id_ch": chunk_id,
                                  "chapter_index": str(i)
                                 })
                ids.append(chunk_id)
                content2index[part.strip()] = i

                new_row = {'id_ch': str(uuid4()),
                           'text_ch': chunks[k],
                           'text_ch_prep': "",
                           'part_index': part_id,
                           'chapter_index': str(i)}
                data_chunks = pd.concat([data_chunks,
                                        pd.DataFrame([new_row], columns=data_chunks.columns)],
                                        ignore_index = True)
                print(f"part len: {len(part)} chunk len: {len(chunks[k])}") # {chunks[k]:50}")

            new_row = {'id_part': part_id, 'text_part': part.strip()}
            data_parts = pd.concat([data_parts,
                                    pd.DataFrame([new_row], columns=data_parts.columns)],
                                   ignore_index = True)
            part_cnt += 1

0
part len: 598 chunk len: 293
part len: 598 chunk len: 291
part len: 598 chunk len: 12
1
part len: 347 chunk len: 297
part len: 347 chunk len: 49
2
part len: 825 chunk len: 286
part len: 825 chunk len: 296
part len: 825 chunk len: 241
3
part len: 1297 chunk len: 300
part len: 1297 chunk len: 293
part len: 1297 chunk len: 299
part len: 1297 chunk len: 299
part len: 1297 chunk len: 102
4 длина: 73
5
part len: 1068 chunk len: 292
part len: 1068 chunk len: 298
part len: 1068 chunk len: 292
part len: 1068 chunk len: 183
6
part len: 2004 chunk len: 291
part len: 2004 chunk len: 285
part len: 2004 chunk len: 298
part len: 2004 chunk len: 290
part len: 2004 chunk len: 294
part len: 2004 chunk len: 287
part len: 2004 chunk len: 253
7
part len: 1177 chunk len: 297
part len: 1177 chunk len: 299
part len: 1177 chunk len: 296
part len: 1177 chunk len: 282
8 длина: 65
9
part len: 1272 chunk len: 286
part len: 1272 chunk len: 298
part len: 1272 chunk len: 298
part len: 1272 chunk len: 293
part len: 

In [48]:
data_chunks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   id_ch          259 non-null    object
 1   text_ch        259 non-null    object
 2   text_ch_prep   259 non-null    object
 3   chapter_index  259 non-null    object
 4   part_index     259 non-null    object
dtypes: object(5)
memory usage: 10.2+ KB


In [49]:
data_parts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46 entries, 0 to 45
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id_part    46 non-null     object
 1   text_part  46 non-null     object
dtypes: object(2)
memory usage: 868.0+ bytes


In [50]:
data_parts.to_csv("parts.csv", index=False)

In [51]:
mystem_analyzer = Mystem()

Installing mystem to /root/.local/bin/mystem from http://download.cdn.yandex.net/mystem/mystem-3.1-linux-64bit.tar.gz


In [52]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

russian_stopwords = stopwords.words('russian')
print(russian_stopwords[:10])

['и', 'в', 'во', 'не', 'что', 'он', 'на', 'я', 'с', 'со']


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [59]:
def chunk_preparation(text: str) -> str:
    """Препроцессинг текста для BM25"""
    if pd.isna(text) or text == "":
        return ""

    prep_text = text.lower()
    prep_text = prep_text.replace("\n", " ")
    prep_text = prep_text.replace("..", ".")
    prep_text = re.sub(r"@\w*", "", prep_text)
    # prep_text = re.sub("[^а-яА-ЯёЁ0-9 ]", " ", prep_text)
    prep_text = mystem_analyzer.lemmatize(prep_text)

    # Фильтрация стоп-слов и пустых токенов
    # prep_text = [word for word in prep_text if word not in russian_stopwords and word.strip()]

    return ' '.join(prep_text)

In [60]:
data_chunks['text_ch_prep'] = data_chunks['text_ch'].apply(chunk_preparation)

In [61]:
data_chunks.head()

id_ch  \
0  16e822c9-cd79-4722-b75e-e7f411c48f70   
1  b87e2f02-2d36-454b-82c1-8d1d59b8ba9a   
2  d605e97a-0f9d-4586-b1f5-fee07db18ac9   
3  e500edde-f5c8-412b-96cb-feef389fe2e0   
4  8583d441-dcb7-413a-baf2-2a9ec868f620   

                                             text_ch  \
0  Программный модуль CAD разработан на базе инте...   
1  2D-представления;задание атрибутивной информац...   
2                                       отображения.   
3  Достаточными условиями выполнения программного...   
4  диске, необходимого для установки – 1 ГБ и более.   

                                        text_ch_prep chapter_index  \
0  программный   модуль   cad   разрабатывать   н...             0   
1  2d - представление ; задание   атрибутивный   ...             0   
2                                   отображение . \n             0   
3  достаточный   условие   выполнение   программн...             1   
4  диск ,  необходимый   для   установка  –  1   ...             1   

                             part_index  
0  2ae8ab96-6cf6-4218-b3d3-cd0386e3618e  
1  2ae8ab96-6cf6-4218-b3d3-cd0386e3618e  
2  2ae8ab96-6cf6-4218-b3d3-cd0386e3618e  
3  464808bc-9a65-4f9f-8d61-78270a8d3a47  
4  464808bc-9a65-4f9f-8d61-78270a8d3a47

In [88]:
data_chunks.tail()

id_ch  \
254  fdcf1dfa-7f50-432b-a8cd-cd490003a1df   
255  10862aee-78a1-4fda-92f8-f6c336dcb590   
256  7b2f367e-2c5f-43e1-b0d2-e1c05f5c301e   
257  11942e46-718a-46b4-8302-9e0dd0a5f5d7   
258  24bf2c45-13a4-4ded-80ed-b19fa8d14172   

                                               text_ch  \
254  представления (2D и 3D) в плоскости экрана, ос...   
255  графическом окне. Изменение направления прокру...   
256  правом верхнем углу рабочего окна (рис. Рисуно...   
257  модель согласно базовому положению. Пиктограмм...   
258  помощью 3Dманипулятора типа SpaceMouse. Для эт...   

                                          text_ch_prep chapter_index  \
254  представление  ( 2d   и   3d )  в   плоскость ...            39   
255  графический   окно .   изменение   направление...            39   
256  правый   верхний   угол   рабочий   окно  ( ри...            39   
257  модель   согласно   базовый   положение .   пи...            39   
258  помощь   3dманипулятора   тип   spacemouse .  ...            39   

                               part_index  
254  8f13e3b2-83e9-4d11-8b8c-6d0fba5145d1  
255  8f13e3b2-83e9-4d11-8b8c-6d0fba5145d1  
256  8f13e3b2-83e9-4d11-8b8c-6d0fba5145d1  
257  8f13e3b2-83e9-4d11-8b8c-6d0fba5145d1  
258  8f13e3b2-83e9-4d11-8b8c-6d0fba5145d1

In [62]:
data_chunks.to_csv("chunks.csv", index=False)

In [63]:
sentence_transformer_ef = embedding_functions.SentenceTransformerEmbeddingFunction(model_name="intfloat/multilingual-e5-large")
sentence_transformer_ef_cl = HuggingFaceEmbeddings(model_name="intfloat/multilingual-e5-large")

modules.json:   0%|          | 0.00/387 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/690 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/418 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/201 [00:00<?, ?B/s]

In [64]:
class SemanticSearchEvaluation:

    def __init__(
        self,
        emb #_path: str

    ):
        # Загружаем эмбеддинги
        # self.emb_path = emb_path

        # Проверка доступности GPU и установка устройства
        # self.device = "cuda" if torch.cuda.is_available() else "cpu"
        # print(f"Using device: {self.device}")

        self.embedding_function = sentence_transformer_ef_cl
        # self.embedding_function = HuggingFaceEmbeddings(
        #     model_name=self.emb_path,
        #     model_kwargs={'device': self.device},
        #     encode_kwargs={
        #         'normalize_embeddings': True,
        #         'convert_to_tensor': True
        #     },
        #     show_progress=False
        # )
        self.result = dict()
        self.collection_name = "data_collection"

    def _get_best_device(self) -> str:
        """Определяет оптимальное устройство для вычислений (GPU/CPU)."""
        if torch.cuda.is_available():
            print(f"GPU доступен: {torch.cuda.get_device_name(0)}")
            return "cuda"
        else:
            print("GPU недоступен, используется CPU")
            return "cpu"

    def evaluate(self,
                 db_path: str,
                 data: pd.DataFrame,
                 column_name: str,
                 displacement: int=0):
        # Загружаем базу Chroma
        # Создаём клиента
        parent_folder = Path(os.getcwd()) # Path(__file__).parent.parent
        path = str(parent_folder / db_path)
        # print(f"path: {path}")

        self.client = PersistentClient(path=path)

        # Инициализация векторного хранилища с embedding функцией
        self.vector_store = Chroma(
            client=self.client,
            collection_name=self.collection_name,
            embedding_function=self.embedding_function,
            collection_metadata={"hnsw:space": "cosine"}
        )

        # Получаем коллекцию для прямых операций
        self.collection = self.vector_store._collection

        # датасет
        df = data

        total = 0    # Вопросов всего
        correct = 0  # Правильных индексов (первый верный)
        both = 0   # Оба индекса верные
        present = 0  # Индекс есть в списке из 5

        for i in tqdm(range(displacement, df.shape[0])):
            if df.iloc[i][column_name] == "-" or df.iloc[i][column_name] != df.iloc[i][column_name]:
                continue

            curr_indexes = []
            vect_res = self.vector_store.similarity_search(
                df.iloc[i][column_name],
                k=5
            )
            len_vec = len(vect_res)
            for j in range(len_vec):
                curr_indexes.append(int(vect_res[j].metadata['index']))
            # print(f"i: {i} curr_indexes: {curr_indexes}")
            total += 1
            if len(curr_indexes) and i == curr_indexes[0]:
                correct += 1
            if len(curr_indexes) and (i == curr_indexes[0] or i == curr_indexes[1]):
                both += 1
            if i in curr_indexes:
                present += 1

        self.result["total"] = total
        self.result["correct"] = correct
        self.result["both"] = both
        self.result["present"] = present
        self.result["top1"] = self.result["correct"]/self.result["total"]*100
        self.result["top2"] = self.result["both"]/self.result["total"]*100
        self.result["top5"] = self.result["present"]/self.result["total"]*100

        return self.result

    def show_result(self):
        print(f'Вопросов всего: {self.result["total"]}')
        print(f'Правильных индексов (первый верный): {self.result["correct"]} доля: {self.result["top1"]:.2f}%')
        print(f'Первый или второй верные: {self.result["both"]} доля: {self.result["top2"]:.2f}%')
        print(f'Индекс есть в списке из 5: {self.result["present"]} доля: {self.result["top5"]:.2f}%')

In [65]:
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device for embedding functions: {device}")

Using device for embedding functions: cpu


In [66]:
persistent_client_serv = chromadb.PersistentClient(path="./db_data")
collection_serv = persistent_client_serv.get_or_create_collection("data_collection", embedding_function=sentence_transformer_ef)

In [67]:
len(documents), len(metadatas), len(ids)

(259, 259, 259)

In [68]:
start = time.time()

collection_serv.add(
    ids=ids,
    documents=documents,
    metadatas=metadatas
)

end = time.time()
print(f"Время исполнения: {(end - start):.3f} с.")

Время исполнения: 258.148 с.


In [69]:
start_time = time.time()

vector_store_serv_from_client = Chroma(
    client=persistent_client_serv,
    collection_name="data_collection",
    embedding_function=sentence_transformer_ef_cl,
)

response_time = time.time() - start_time
print(f"Время построения клиента: {response_time:.3f} секунды")

Время построения клиента: 0.003 секунды


In [70]:
question = "Программный модуль"

In [71]:
vect_res = vector_store_serv_from_client.similarity_search(
    question,
    k=5
)

In [72]:
for i in range(len(vect_res)):
    print(f"{vect_res[i].page_content}\n")

1 Назначение программного модуля CAD
отображения.

2 Условия выполнения программного модуля CAD
диске, необходимого для установки – 1 ГБ и более.

3.2 Управление приложениями
Данное окно предназначено для осуществления управления дополнительными программными компонентами. Вызов окна осуществляется с помощью пиктограммы «Приложения»  из панели «Лента» (вкладка «Вид», группа «Настройки»). Также данное окно может быть вызвано из выпадающего меню настроек программного модуля

3.2 Управление приложениями
помощью кнопки «Добавить», при нажатии которой на рабочем столе будет отображен стандартный диалог операционной системы «Открыть».Отключение модуля осуществляется после его выбора в перечне подключенных с помощью кнопки «Удалить».Изменение состояния модулей происходит после перезагрузки системы.Для

4.1 Окна программного модуля CAD
окон не меняет функциональность работы программного модуля CAD в текущем сеансе. Программный модуль CAD завершает работу при закрытии последнего главного окна.От

In [73]:
docs = []
for i in range(len(vect_res)):
    if vect_res[i].metadata["content"] not in docs:
        docs.append(vect_res[i].metadata["content"])
        print(f"{i+1} {vect_res[i].metadata["content"]}\n")

1 Программный модуль CAD разработан на базе интегрированной инженерной программной платформы и программно-математического ядра трехмерного моделирования «RGK».Функциональное назначение программного модуля CAD:создание и редактирование геометрической модели объектов проектирования, включая 3D- и 2D-представления;задание атрибутивной информации, связанной с точностными, размерными, технологическими параметрами объектов проектирования (PMI);оформление графической документации в соответствии с требованиями стандартов (ЕСКД);визуализация объектов проектирования с различными параметрами отображения.

2 Достаточными условиями выполнения программного модуля CAD являются системные требования:ОС – Microsoft Windows версии не ниже 10, Astra Linux версии 1.7 и выше;процессор Intel Core i3 или выше;видеокарта NVIDIA или AMD с памятью 1ГБ и выше;ОЗУ 2 ГБ и более;объем свободного пространства на жестком диске, необходимого для установки – 1 ГБ и более.

3 Данное окно предназначено для осуществления у

In [74]:
retriver_test = SemanticSearchEvaluation(sentence_transformer_ef_cl) #"intfloat/multilingual-e5-large")

res = retriver_test.evaluate("db_data", data, "question")
res

100%|██████████| 40/40 [00:30<00:00,  1.33it/s]


{'total': 40,
 'correct': 37,
 'both': 39,
 'present': 40,
 'top1': 92.5,
 'top2': 97.5,
 'top5': 100.0}

### BM25

In [76]:
from rank_bm25 import BM25Okapi

In [77]:
tokenized_corpus = [doc.split(" ") for doc in data_chunks['text_ch_prep'].tolist()]

bm25 = BM25Okapi(tokenized_corpus)

In [78]:
data.question[0]

'В чём состоит функциональное назначение CAD‑модуля, построенного на единой инженерной платформе с ядром трёхмерного моделирования RGK, относительно создания и изменения геометрических моделей (включая 3‑D и плоские представления), задания атрибутов PMI, подготовки чертежей согласно требованиям ЕСКД и визуализации с различными параметрами отображения?'

In [79]:
chunk_preparation(data.question[0])

'в   что   состоять   функциональный   назначение   cad ‑ модуль ,  построить   на   единый   инженерный   платформа   с   ядро   трехмерный   моделирование   rgk ,  относительно   создание   и   изменение   геометрический   модель  ( включая   3 ‑ d   и   плоский   представление ),  задание   атрибут   pmi ,  подготовка   чертеж   согласно   требование   ескд   и   визуализация   с   различный   параметр   отображение ? \n'

In [80]:
def bm25_doc_scores(query, top_k=10):
    scores = bm25.get_scores(chunk_preparation(query))
    # print(f"scores: {scores}")
    top_idx = np.argsort(scores)[::-1][:top_k]
    # print(f"top_idx: {top_idx}")
    return {int(i): float(scores[i]) for i in top_idx}, {float(scores[i]): int(i) for i in top_idx}

In [81]:
data.question[5]

'Каким способом можно включить пользовательский шаблон в перечень, представленный диалогом выбора прототипа, и какие конфигурационные файлы при этом требуется изменить?'

In [82]:
i2s, s2i = bm25_doc_scores(data.question[5])
i2s, s2i

({104: 116.23166670833402,
  192: 85.9014785025612,
  94: 82.94648028530106,
  129: 81.82681177021352,
  111: 80.2785573315342,
  175: 80.2090044897064,
  77: 79.41055629525678,
  228: 68.56383560199937,
  17: 68.32947793913901,
  44: 63.512842799311564},
 {116.23166670833402: 104,
  85.9014785025612: 192,
  82.94648028530106: 94,
  81.82681177021352: 129,
  80.2785573315342: 111,
  80.2090044897064: 175,
  79.41055629525678: 77,
  68.56383560199937: 228,
  68.32947793913901: 17,
  63.512842799311564: 44})

In [83]:
max(s2i.keys()), s2i[max(s2i.keys())], int(data_chunks.chapter_index[s2i[max(s2i.keys())]])

(116.23166670833402, 104, 25)

In [96]:
class BM25SearchEvaluation:

    def __init__(
        self,
        bm25

    ):
        self.bm25 = bm25
        self.result = dict()

    def evaluate(self,
                 data: pd.DataFrame,
                 column_name: str,
                 displacement: int=0):

        # датасет
        df = data

        total = 0    # Вопросов всего
        correct = 0  # Правильных индексов (первый верный)
        both = 0   # Оба индекса верные
        present = 0  # Индекс есть в списке из 5

        for i in tqdm(range(displacement, df.shape[0])):
            if df.iloc[i][column_name] == "-" or df.iloc[i][column_name] != df.iloc[i][column_name]:
                continue

            curr_indexes = []

            # print(df.iloc[i][column_name])
            i2s, s2i = bm25_doc_scores(df.iloc[i][column_name], top_k=5)
            # print(f"i2s: {i2s}")

            for k, v in s2i.items():
                curr_indexes.append(int(data_chunks.chapter_index[v]))

            print(f"Индекс вопроса: {i} найденный индекс: {curr_indexes}")
            total += 1
            if len(curr_indexes) and i == curr_indexes[0]:
                correct += 1
            if len(curr_indexes) and (i == curr_indexes[0] or i == curr_indexes[1]):
                both += 1
            if i in curr_indexes:
                present += 1

        self.result["total"] = total
        self.result["correct"] = correct
        self.result["both"] = both
        self.result["present"] = present
        self.result["top1"] = self.result["correct"]/self.result["total"]*100
        self.result["top2"] = self.result["both"]/self.result["total"]*100
        self.result["top5"] = self.result["present"]/self.result["total"]*100

        return self.result

    def show_result(self):
        print(f'Вопросов всего: {self.result["total"]}')
        print(f'Правильных индексов (первый верный): {self.result["correct"]} доля: {self.result["top1"]:.2f}%')
        print(f'Первый или второй верные: {self.result["both"]} доля: {self.result["top2"]:.2f}%')
        print(f'Индекс есть в списке из 5: {self.result["present"]} доля: {self.result["top5"]:.2f}%')

In [97]:
bm25_evaluation = BM25SearchEvaluation(bm25)

res = bm25_evaluation.evaluate(data, "question")
res

 20%|██        | 8/40 [00:00<00:00, 33.71it/s]

Индекс вопроса: 0 найденный индекс: [25, 27, 16, 33, 26]
Индекс вопроса: 1 найденный индекс: [25, 33, 30, 26, 27]
Индекс вопроса: 2 найденный индекс: [25, 33, 27, 16, 30]
Индекс вопроса: 3 найденный индекс: [25, 33, 27, 16, 21]
Индекс вопроса: 4 найденный индекс: [25, 33, 16, 27, 21]
Индекс вопроса: 5 найденный индекс: [25, 33, 21, 27, 26]
Индекс вопроса: 6 найденный индекс: [25, 27, 16, 33, 21]
Индекс вопроса: 7 найденный индекс: [25, 33, 27, 16, 30]


 40%|████      | 16/40 [00:00<00:00, 34.32it/s]

Индекс вопроса: 8 найденный индекс: [25, 26, 30, 35, 33]
Индекс вопроса: 9 найденный индекс: [25, 27, 33, 16, 30]
Индекс вопроса: 10 найденный индекс: [25, 33, 27, 16, 21]
Индекс вопроса: 11 найденный индекс: [25, 27, 33, 16, 30]
Индекс вопроса: 12 найденный индекс: [25, 33, 27, 16, 26]
Индекс вопроса: 13 найденный индекс: [25, 26, 30, 33, 21]
Индекс вопроса: 14 найденный индекс: [25, 30, 21, 33, 26]
Индекс вопроса: 15 найденный индекс: [25, 27, 26, 16, 30]


 60%|██████    | 24/40 [00:00<00:00, 34.20it/s]

Индекс вопроса: 16 найденный индекс: [25, 33, 21, 30, 26]
Индекс вопроса: 17 найденный индекс: [25, 27, 16, 30, 26]
Индекс вопроса: 18 найденный индекс: [25, 33, 27, 16, 26]
Индекс вопроса: 19 найденный индекс: [25, 27, 16, 33, 30]
Индекс вопроса: 20 найденный индекс: [25, 27, 16, 33, 30]
Индекс вопроса: 21 найденный индекс: [25, 33, 30, 26, 27]
Индекс вопроса: 22 найденный индекс: [25, 33, 27, 16, 30]
Индекс вопроса: 23 найденный индекс: [25, 27, 16, 26, 30]


 70%|███████   | 28/40 [00:00<00:00, 33.15it/s]

Индекс вопроса: 24 найденный индекс: [25, 27, 16, 33, 30]
Индекс вопроса: 25 найденный индекс: [25, 27, 16, 33, 30]
Индекс вопроса: 26 найденный индекс: [25, 27, 16, 33, 21]
Индекс вопроса: 27 найденный индекс: [25, 27, 16, 33, 21]
Индекс вопроса: 28 найденный индекс: [25, 33, 27, 16, 26]
Индекс вопроса: 29 найденный индекс: [25, 33, 27, 16, 30]
Индекс вопроса: 30 найденный индекс: [25, 33, 27, 16, 21]


 90%|█████████ | 36/40 [00:01<00:00, 30.57it/s]

Индекс вопроса: 31 найденный индекс: [25, 27, 16, 33, 21]
Индекс вопроса: 32 найденный индекс: [25, 33, 30, 26, 27]
Индекс вопроса: 33 найденный индекс: [25, 27, 33, 16, 26]
Индекс вопроса: 34 найденный индекс: [25, 27, 16, 33, 26]
Индекс вопроса: 35 найденный индекс: [25, 27, 16, 21, 33]
Индекс вопроса: 36 найденный индекс: [25, 21, 30, 33, 26]
Индекс вопроса: 37 найденный индекс: [25, 27, 33, 16, 30]


100%|██████████| 40/40 [00:01<00:00, 31.68it/s]

Индекс вопроса: 38 найденный индекс: [25, 33, 27, 16, 26]
Индекс вопроса: 39 найденный индекс: [25, 27, 16, 33, 30]


{'total': 40,
 'correct': 1,
 'both': 2,
 'present': 3,
 'top1': 2.5,
 'top2': 5.0,
 'top5': 7.5}

In [98]:
bm25_evaluation.show_result()

Вопросов всего: 40
Правильных индексов (первый верный): 1 доля: 2.50%
Первый или второй верные: 2 доля: 5.00%
Индекс есть в списке из 5: 3 доля: 7.50%


Для каждой главы сгенерирован ворос. Каждой главе соответствует строка в таблице data. Соответственно корректным значением индекса при поиске является значение индекса совпадающее с индексом строки вопроса. Стандартный векторный поиск продемонстрировал хороший результат: top1: 92.5%, top2: 97.5%, top5: 1000%. Где top1 это совпадение в певой позиции top5 совпадение в первых пяти позициях. Поиск BM25 не работает корректно на данном корпусе и я даже подозреваю ошибку в реализации.

# Задание 3 Добавить к системе реранкер

Замерить улучшение/ухудшение качества ретрива, замерить время, требуемое на запрос и его изменение

In [125]:
data = pd.read_csv("parsed_qa.csv")
data.head()

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   

                                            question  
0  В чём состоит функциональное назначение CAD‑мо...  
1  Какие минимальные программные и аппаратные хар...  
2  Каким способом пользователь может открыть прив...  
3  Каким образом можно вызвать диалог управления ...  
4  В каком пункте меню отображается список средст...

In [126]:
JINA_API_KEY = userdata.get('JINA_API_KEY')

In [92]:
import requests

def jina_rerank_api(query, documents, top_n=3):
    api_key = JINA_API_KEY

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}"
    }

    data_send = {
        "model": "jina-reranker-v2-base-multilingual",
        "query": query,
        "documents": documents,
        "top_n": top_n
    }

    response = requests.post(
        "https://api.jina.ai/v1/rerank",
        headers=headers,
        json=data_send
    )

    if response.status_code == 200:
        return response.json()["results"]
    else:
        raise Exception(f"API Error: {response.status_code}")

# Пример использования с русскоязычным контентом
russian_query = "методы обработки естественного языка"
russian_documents = [
    "Глубокое обучение для обработки русского языка",
    "Статистические методы анализа текстов",
    "Трансформеры в машинном переводе для русского языка",
    "Семантический поиск на русском языке",
    "Морфологический анализ русских текстов",
    "Построение эмбеддингов Wor2vec и TF-IDF"
]

results = jina_rerank_api(russian_query, russian_documents)
print(results)
for result in results:
    print(f"Документ: {result['document']}")
    print(f"Релевантность: {result['relevance_score']:.4f}")

[{'index': 0, 'relevance_score': 0.20307463, 'document': 'Глубокое обучение для обработки русского языка'}, {'index': 5, 'relevance_score': 0.16238002, 'document': 'Построение эмбеддингов Wor2vec и TF-IDF'}, {'index': 1, 'relevance_score': 0.15507847, 'document': 'Статистические методы анализа текстов'}]
Документ: Глубокое обучение для обработки русского языка
Релевантность: 0.2031
Документ: Построение эмбеддингов Wor2vec и TF-IDF
Релевантность: 0.1624
Документ: Статистические методы анализа текстов
Релевантность: 0.1551


In [112]:
question = "Программный модуль"

vect_res = vector_store_serv_from_client.similarity_search(
    question,
    k=20
)

In [113]:
vect_res[0].metadata['content']

'Программный модуль CAD разработан на базе интегрированной инженерной программной платформы и программно-математического ядра трехмерного моделирования «RGK».Функциональное назначение программного модуля CAD:создание и редактирование геометрической модели объектов проектирования, включая 3D- и 2D-представления;задание атрибутивной информации, связанной с точностными, размерными, технологическими параметрами объектов проектирования (PMI);оформление графической документации в соответствии с требованиями стандартов (ЕСКД);визуализация объектов проектирования с различными параметрами отображения.'

In [122]:
text2index = {}
for i, text in enumerate(data.text):
    text2index[text] = i

In [115]:
candidates = []

for candidate in vect_res:
    if candidate.metadata['content'] not in candidates:
        candidates.append(candidate.metadata['content'])
    if len(candidates) == 8:
        break

len(candidates)

8

In [124]:
headers = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {JINA_API_KEY}"
}

data_send = {
    "model": "jina-reranker-v2-base-multilingual",
    "query": question,
    "documents": candidates,
    "top_n": 5,
    "return_documents": True
}

# Выполнение запроса
response = requests.post(
    "https://api.jina.ai/v1/rerank",
    headers=headers,
    json=data_send
)

if response.status_code == 200:
    rerank_results = response.json()
    print("Реранкинг завершен. Топ-5 документов:")
    print("=" * 50)

    # Вывод переранжированных результатов
    for idx, result in enumerate(rerank_results['results']):
        print(f"{idx+1}. [Релевантность: {result['relevance_score']:.4f} Индекс: {text2index[result['document']]}]\n{result['document']}\n")
        print("-" * 50)

else:
    print(f"Ошибка API: {response.status_code}")
    print(response.text)

Реранкинг завершен. Топ-5 документов:
1. [Релевантность: 0.5974 Индекс: 0]
Программный модуль CAD разработан на базе интегрированной инженерной программной платформы и программно-математического ядра трехмерного моделирования «RGK».Функциональное назначение программного модуля CAD:создание и редактирование геометрической модели объектов проектирования, включая 3D- и 2D-представления;задание атрибутивной информации, связанной с точностными, размерными, технологическими параметрами объектов проектирования (PMI);оформление графической документации в соответствии с требованиями стандартов (ЕСКД);визуализация объектов проектирования с различными параметрами отображения.

--------------------------------------------------
2. [Релевантность: 0.5482 Индекс: 15]
Программный модуль CAD поддерживает многодокументный многооконный интерфейс (MDI). Это позволяет создавать произвольное число главных окон, каждое из которых может размещаться произвольно, в том числе на разных мониторах.Главные окна яв

In [144]:
class RerankedSearchEvaluation:

    def __init__(
        self
    ):
        self.result = dict()

    def evaluate(self,
                 data: pd.DataFrame,
                 column_name: str,
                 displacement: int=0):

        # датасет
        df = data

        total = 0    # Вопросов всего
        correct = 0  # Правильных индексов (первый верный)
        both = 0   # Оба индекса верные
        present = 0  # Индекс есть в списке из 5

        for i in range(displacement, df.shape[0]): #tqdm(range(displacement, df.shape[0])):
            if df.iloc[i][column_name] == "-" or df.iloc[i][column_name] != df.iloc[i][column_name]:
                continue

            curr_indexes = []

            # print(df.iloc[i][column_name])
            vect_res = vector_store_serv_from_client.similarity_search(
                df.iloc[i][column_name],
                k=20
            )

            candidates = []
            indexes_before = []

            for candidate in vect_res:
                if candidate.metadata['content'] not in candidates:
                    candidates.append(candidate.metadata['content'])
                    indexes_before.append(int(candidate.metadata['index']))
                if len(candidates) == 5: #8:
                    break


            headers = {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {JINA_API_KEY}"
            }

            data_send = {
                "model": "jina-reranker-v2-base-multilingual",
                "query": df.iloc[i][column_name],
                "documents": candidates,
                "top_n": 5,
                "return_documents": True
            }

            # Выполнение запроса
            response = requests.post(
                "https://api.jina.ai/v1/rerank",
                headers=headers,
                json=data_send
            )

            if response.status_code == 200:
                rerank_results = response.json()
                # print("Реранкинг завершен. Топ-5 документов:")
                # print("=" * 50)

                # Вывод переранжированных результатов
                for idx, result in enumerate(rerank_results['results']):
                    # print(f"{idx+1}. [Релевантность: {result['relevance_score']:.4f} Индекс: {text2index[result['document']]}]\n{result['document']}\n")
                    # print("-" * 50)
                    curr_indexes.append(int(content2index[result['document']]))

            else:
                print(f"Ошибка API: {response.status_code}")
                print(response.text)
            # print(f"Индексы до реранкинга: {indexes_before} индексы после реранкинга: {curr_indexes}")
            print(f"{indexes_before} до")
            print(f"{curr_indexes} после")

            print(f"{i} (индекс вопроса) найденный индекс: {curr_indexes}\n")
            total += 1
            if len(curr_indexes) and i == curr_indexes[0]:
                correct += 1
            if len(curr_indexes) and (i == curr_indexes[0] or i == curr_indexes[1]):
                both += 1
            if i in curr_indexes:
                present += 1

        self.result["total"] = total
        self.result["correct"] = correct
        self.result["both"] = both
        self.result["present"] = present
        self.result["top1"] = self.result["correct"]/self.result["total"]*100
        self.result["top2"] = self.result["both"]/self.result["total"]*100
        self.result["top5"] = self.result["present"]/self.result["total"]*100

        return self.result

    def show_result(self):
        print(f'Вопросов всего: {self.result["total"]}')
        print(f'Правильных индексов (первый верный): {self.result["correct"]} доля: {self.result["top1"]:.2f}%')
        print(f'Первый или второй верные: {self.result["both"]} доля: {self.result["top2"]:.2f}%')
        print(f'Индекс есть в списке из 5: {self.result["present"]} доля: {self.result["top5"]:.2f}%')

#### Реранкинг 8 кандидатов

In [142]:
reranked_evaluation = RerankedSearchEvaluation()

res = reranked_evaluation.evaluate(data, "question")
res

[0, 29, 2, 16, 15, 6, 5, 14] до
[0, 16, 29, 15, 5] после
0 (индекс вопроса) найденный индекс: [0, 16, 29, 15, 5]

[1, 2, 23, 14, 15, 18, 28, 28] до
[1, 2, 28, 28, 14] после
1 (индекс вопроса) найденный индекс: [1, 2, 28, 28, 14]

[2, 14, 18, 3, 17, 21, 27, 20] до
[2, 14, 21, 18, 17] после
2 (индекс вопроса) найденный индекс: [2, 14, 21, 18, 17]

[3, 17, 14, 2, 27, 21, 26, 39] до
[3, 21, 27, 17, 26] после
3 (индекс вопроса) найденный индекс: [3, 21, 27, 17, 26]

[4, 22, 14, 9, 10, 6, 24, 25] до
[4, 22, 14, 10, 6] после
4 (индекс вопроса) найденный индекс: [4, 22, 14, 10, 6]

[5, 26, 27, 6, 28, 14, 2, 13] до
[5, 14, 2, 28, 6] после
5 (индекс вопроса) найденный индекс: [5, 14, 2, 28, 6]

[6, 33, 33, 28, 2, 33, 39, 30] до
[6, 28, 33, 33, 2] после
6 (индекс вопроса) найденный индекс: [6, 28, 33, 33, 2]

[7, 28, 28, 33, 14, 33, 5, 3] до
[7, 28, 33, 28, 33] после
7 (индекс вопроса) найденный индекс: [7, 28, 33, 28, 33]

[8, 16, 10, 9, 22, 17, 6, 19] до
[8, 10, 9, 16, 17] после
8 (индекс вопро

{'total': 40,
 'correct': 38,
 'both': 39,
 'present': 40,
 'top1': 95.0,
 'top2': 97.5,
 'top5': 100.0}

In [143]:
reranked_evaluation.show_result()

Вопросов всего: 40
Правильных индексов (первый верный): 38 доля: 95.00%
Первый или второй верные: 39 доля: 97.50%
Индекс есть в списке из 5: 40 доля: 100.00%


#### Реранкинг 5 кандидатов

In [145]:
reranked_evaluation = RerankedSearchEvaluation()

res = reranked_evaluation.evaluate(data, "question")
res

[0, 29, 2, 16, 15] до
[0, 16, 29, 15, 2] после
0 (индекс вопроса) найденный индекс: [0, 16, 29, 15, 2]

[1, 2, 23, 14, 15] до
[1, 2, 14, 15, 23] после
1 (индекс вопроса) найденный индекс: [1, 2, 14, 15, 23]

[2, 14, 18, 3, 17] до
[2, 14, 18, 17, 3] после
2 (индекс вопроса) найденный индекс: [2, 14, 18, 17, 3]

[3, 17, 14, 2, 27] до
[3, 27, 17, 14, 2] после
3 (индекс вопроса) найденный индекс: [3, 27, 17, 14, 2]

[4, 22, 14, 9, 10] до
[4, 22, 14, 10, 9] после
4 (индекс вопроса) найденный индекс: [4, 22, 14, 10, 9]

[5, 26, 27, 6, 28] до
[5, 28, 6, 27, 26] после
5 (индекс вопроса) найденный индекс: [5, 28, 6, 27, 26]

[6, 33, 33, 28, 2] до
[6, 28, 33, 33, 2] после
6 (индекс вопроса) найденный индекс: [6, 28, 33, 33, 2]

[7, 28, 28, 33, 14] до
[7, 28, 33, 28, 14] после
7 (индекс вопроса) найденный индекс: [7, 28, 33, 28, 14]

[8, 16, 10, 9, 22] до
[8, 10, 9, 16, 22] после
8 (индекс вопроса) найденный индекс: [8, 10, 9, 16, 22]

[9, 34, 28, 12, 28] до
[9, 34, 28, 28, 12] после
9 (индекс во

{'total': 40,
 'correct': 38,
 'both': 40,
 'present': 40,
 'top1': 95.0,
 'top2': 100.0,
 'top5': 100.0}

In [146]:
reranked_evaluation.show_result()

Вопросов всего: 40
Правильных индексов (первый верный): 38 доля: 95.00%
Первый или второй верные: 40 доля: 100.00%
Индекс есть в списке из 5: 40 доля: 100.00%


Реранкинг позволил повысить качество извлечения (вполне впечатляет)

# Задание 4/5 Сгенерировать ответы модели на вопросы из золотого набора и Протестировать оценку LLM-кой релевантности и выбора контекстов

Замерить улучшение/ухудшение качества ретрива, замерить время, требуемое на запрос и его изменение

Разметить сгенерированные данные — агрументировать метод разметки, если это LLM-as-a-judge, обосновать надёжность конкретно вашего промпта и вашего решения разметчика

In [154]:
data['golden_answer'] = ""

In [155]:
llm.set_system_prompt(
            "Ты участвуешь в решении задачи оценки конвейра RAG. "
            "Ты создаёшь эталонные - правильные оветы на основании заведомо правильного контекста"
            "Выводи только текст ответа. Не добавляй ничего лишнего!"
        )

In [156]:
for i in tqdm(range(data.shape[0])):
    if data.iloc[i]["text"]:
        prompt = (
            f"Представь себя пользователем RAG системы. "
            f"Напиши эталонный ответ на основании заведомо правильного контекста."
            f"Выводи только текст ответа. Не добавляй ничего лишнего!"
            f"\nВопрос:\n\n{data.iloc[i]['question']}"
            f"\nКонтекст:\n\n{data.iloc[i]['text']}"
        )

        data.loc[i, 'golden_answer'] = llm.generate_answer(prompt)

100%|██████████| 40/40 [01:27<00:00,  2.19s/it]


In [157]:
data.head(10)

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   
5               3.3.1 Создание нового документа   
6                     3.3.2 Открытие документов   
7                   3.3.3 Сохранение документов   
8                      3.3.4 Закрытие документа   
9                    3.3.5.1 Свойства документа   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   
5  Команда «Новый документ»  предназначена для со...   
6  Команда «Открыть»  вызывает стандартный диалог...   
7  Команда «Сохранить»  выполняет сохранение теку...   
8  Для закрытия документа нажмите Х на вкладке ря...   
9  Свойства активного документа отображаются во в...   

                                            question  \
0  В чём состоит функциональное назначение CAD‑мо...   
1  Какие минимальные программные и аппаратные хар...   
2  Каким способом пользователь может открыть прив...   
3  Каким образом можно вызвать диалог управления ...   
4  В каком пункте меню отображается список средст...   
5  Каким способом можно включить пользовательский...   
6  Каким способом можно запустить открытие RGP‑фа...   
7  Как установить интервал автосохранения и гаран...   
8  Каким образом можно завершить работу с текущим...   
9  Каким способом в окне свойств текущего файла м...   

                                       golden_answer  
0  Функциональное назначение CAD‑модуля, построен...  
1  Минимальные требования для установки и работы ...  
2  Пользователь может открыть приветственную пане...  
3  Если пункт **«Приложения»** в ленте недоступен...  
4                              В пункте меню «Файл».  
5  Чтобы пользовательский шаблон появился в диало...  
6  Запустить открытие RGP‑файла через консоль мож...  
7  Чтобы задать интервал автосохранения и обеспеч...  
8       Нажмите «X» на вкладке рядом с именем файла.  
9  В окне свойств в блоке «Настройки» разверните ...

In [160]:
data['system_answer'] = ""

In [161]:
llm.set_system_prompt(
            "Ты участвуешь в решении задачи оценки конвейра RAG. "
            "Ты создаёшь оветы на основании контекста"
            "Выводи только текст ответа. Не добавляй ничего лишнего!"
        )

In [162]:
for i in tqdm(range(data.shape[0])):
    if data.iloc[i]["text"]:

        df = data
        curr_indexes = []

        vect_res = vector_store_serv_from_client.similarity_search(
            df.iloc[i]['question'],
            k=20
        )

        candidates = []
        indexes_before = []

        for candidate in vect_res:
            if candidate.metadata['content'] not in candidates:
                candidates.append(candidate.metadata['content'])
                indexes_before.append(int(candidate.metadata['index']))
            if len(candidates) == 5: #8:
                break

        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {JINA_API_KEY}"
        }

        data_send = {
            "model": "jina-reranker-v2-base-multilingual",
            "query": df.iloc[i]['question'],
            "documents": candidates,
            "top_n": 5,
            "return_documents": True
        }

        # Выполнение запроса
        response = requests.post(
            "https://api.jina.ai/v1/rerank",
            headers=headers,
            json=data_send
        )

        if response.status_code == 200:
            rerank_results = response.json()

        prompt = (
            f"Представь себя пользователем RAG системы. "
            f"Напиши ответ на основании контекста."
            f"Выводи только текст ответа. Не добавляй ничего лишнего!"
            f"\nВопрос:\n\n{data.iloc[i]['question']}"
            f"\n\nКонтекст:\n\n{rerank_results['results'][0]['document']}"
        )

        # print(prompt)
        # break

        data.loc[i, 'system_answer'] = llm.generate_answer(prompt)

100%|██████████| 40/40 [01:59<00:00,  3.00s/it]


In [163]:
data.head(10)

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   
5               3.3.1 Создание нового документа   
6                     3.3.2 Открытие документов   
7                   3.3.3 Сохранение документов   
8                      3.3.4 Закрытие документа   
9                    3.3.5.1 Свойства документа   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   
5  Команда «Новый документ»  предназначена для со...   
6  Команда «Открыть»  вызывает стандартный диалог...   
7  Команда «Сохранить»  выполняет сохранение теку...   
8  Для закрытия документа нажмите Х на вкладке ря...   
9  Свойства активного документа отображаются во в...   

                                            question  \
0  В чём состоит функциональное назначение CAD‑мо...   
1  Какие минимальные программные и аппаратные хар...   
2  Каким способом пользователь может открыть прив...   
3  Каким образом можно вызвать диалог управления ...   
4  В каком пункте меню отображается список средст...   
5  Каким способом можно включить пользовательский...   
6  Каким способом можно запустить открытие RGP‑фа...   
7  Как установить интервал автосохранения и гаран...   
8  Каким образом можно завершить работу с текущим...   
9  Каким способом в окне свойств текущего файла м...   

                                       golden_answer  \
0  Функциональное назначение CAD‑модуля, построен...   
1  Минимальные требования для установки и работы ...   
2  Пользователь может открыть приветственную пане...   
3  Если пункт **«Приложения»** в ленте недоступен...   
4                              В пункте меню «Файл».   
5  Чтобы пользовательский шаблон появился в диало...   
6  Запустить открытие RGP‑файла через консоль мож...   
7  Чтобы задать интервал автосохранения и обеспеч...   
8       Нажмите «X» на вкладке рядом с именем файла.   
9  В окне свойств в блоке «Настройки» разверните ...   

                                       system_answer  
0  Функциональное назначение CAD‑модуля — обеспеч...  
1  Минимальные требования для установки и работы ...  
2  Пользователь может открыть приветственную пане...  
3  Диалог управления дополнительными модулями CAD...  
4                              В пункте меню «Файл».  
5  Чтобы пользовательский шаблон появился в диало...  
6  Запустить открытие RGP‑файла через консоль мож...  
7  Откройте диалог «Параметры автосохранения». Ус...  
8         Нажмите X на вкладке рядом с именем файла.  
9  В окне свойств в группе «Настройки» раскройте ...

In [166]:
data['retrived_context'] = ""

In [167]:
for i in tqdm(range(data.shape[0])):
    if data.iloc[i]["text"]:

        df = data
        curr_indexes = []

        vect_res = vector_store_serv_from_client.similarity_search(
            df.iloc[i]['question'],
            k=20
        )

        candidates = []
        indexes_before = []

        for candidate in vect_res:
            if candidate.metadata['content'] not in candidates:
                candidates.append(candidate.metadata['content'])
                indexes_before.append(int(candidate.metadata['index']))
            if len(candidates) == 5: #8:
                break

        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {JINA_API_KEY}"
        }

        data_send = {
            "model": "jina-reranker-v2-base-multilingual",
            "query": df.iloc[i]['question'],
            "documents": candidates,
            "top_n": 5,
            "return_documents": True
        }

        # Выполнение запроса
        response = requests.post(
            "https://api.jina.ai/v1/rerank",
            headers=headers,
            json=data_send
        )

        if response.status_code == 200:
            rerank_results = response.json()

        data.loc[i, 'retrived_context'] = rerank_results['results'][0]['document']

100%|██████████| 40/40 [00:47<00:00,  1.18s/it]


In [168]:
data.head(10)

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   
5               3.3.1 Создание нового документа   
6                     3.3.2 Открытие документов   
7                   3.3.3 Сохранение документов   
8                      3.3.4 Закрытие документа   
9                    3.3.5.1 Свойства документа   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   
5  Команда «Новый документ»  предназначена для со...   
6  Команда «Открыть»  вызывает стандартный диалог...   
7  Команда «Сохранить»  выполняет сохранение теку...   
8  Для закрытия документа нажмите Х на вкладке ря...   
9  Свойства активного документа отображаются во в...   

                                            question  \
0  В чём состоит функциональное назначение CAD‑мо...   
1  Какие минимальные программные и аппаратные хар...   
2  Каким способом пользователь может открыть прив...   
3  Каким образом можно вызвать диалог управления ...   
4  В каком пункте меню отображается список средст...   
5  Каким способом можно включить пользовательский...   
6  Каким способом можно запустить открытие RGP‑фа...   
7  Как установить интервал автосохранения и гаран...   
8  Каким образом можно завершить работу с текущим...   
9  Каким способом в окне свойств текущего файла м...   

                                       golden_answer  \
0  Функциональное назначение CAD‑модуля, построен...   
1  Минимальные требования для установки и работы ...   
2  Пользователь может открыть приветственную пане...   
3  Если пункт **«Приложения»** в ленте недоступен...   
4                              В пункте меню «Файл».   
5  Чтобы пользовательский шаблон появился в диало...   
6  Запустить открытие RGP‑файла через консоль мож...   
7  Чтобы задать интервал автосохранения и обеспеч...   
8       Нажмите «X» на вкладке рядом с именем файла.   
9  В окне свойств в блоке «Настройки» разверните ...   

                                       system_answer  \
0  Функциональное назначение CAD‑модуля — обеспеч...   
1  Минимальные требования для установки и работы ...   
2  Пользователь может открыть приветственную пане...   
3  Диалог управления дополнительными модулями CAD...   
4                              В пункте меню «Файл».   
5  Чтобы пользовательский шаблон появился в диало...   
6  Запустить открытие RGP‑файла через консоль мож...   
7  Откройте диалог «Параметры автосохранения». Ус...   
8         Нажмите X на вкладке рядом с именем файла.   
9  В окне свойств в группе «Настройки» раскройте ...   

                                    retrived_context  
0  Программный модуль CAD разработан на базе инте...  
1  Достаточными условиями выполнения программного...  
2  Загрузка и запуск программного модуля CAD осущ...  
3  Данное окно предназначено для осуществления уп...  
4  Перечень инструментов для работы с документами...  
5  Команда «Новый документ»  предназначена для со...  
6  Команда «Открыть»  вызывает стандартный диалог...  
7  Команда «Сохранить»  выполняет сохранение теку...  
8  Для закрытия документа нажмите Х на вкладке ря...  
9  Свойства активного документа отображаются во в...

In [169]:
data.to_csv("parsed_qa.csv", index=False)

In [170]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   header            40 non-null     object
 1   text              40 non-null     object
 2   question          40 non-null     object
 3   golden_answer     40 non-null     object
 4   system_answer     40 non-null     object
 5   retrived_context  40 non-null     object
dtypes: object(6)
memory usage: 2.0+ KB


### Оценка с использованием RAGAS

In [28]:
import pandas as pd
import numpy as np
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import (
    faithfulness,
    answer_relevancy,
    context_precision,
    context_recall,
    answer_correctness,
    answer_similarity
)
from ragas.run_config import RunConfig

# Импорты для BLEU и ROUGE
import nltk
nltk.download('punkt', quiet=True)
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge_score import rouge_scorer

from langchain_openai import ChatOpenAI, OpenAIEmbeddings
from langchain.embeddings import HuggingFaceEmbeddings

In [3]:
data = pd.read_csv("parsed_qa.csv")
data.head()

header  \
0          1 Назначение программного модуля CAD   
1  2 Условия выполнения программного модуля CAD   
2            3.1 Запуск программного модуля CAD   
3                   3.2 Управление приложениями   
4                      3.3 Работа с документами   

                                                text  \
0  Программный модуль CAD разработан на базе инте...   
1  Достаточными условиями выполнения программного...   
2  Загрузка и запуск программного модуля CAD осущ...   
3  Данное окно предназначено для осуществления уп...   
4  Перечень инструментов для работы с документами...   

                                            question  \
0  В чём состоит функциональное назначение CAD‑мо...   
1  Какие минимальные программные и аппаратные хар...   
2  Каким способом пользователь может открыть прив...   
3  Каким образом можно вызвать диалог управления ...   
4  В каком пункте меню отображается список средст...   

                                       golden_answer  \
0  Функциональное назначение CAD‑модуля, построен...   
1  Минимальные требования для установки и работы ...   
2  Пользователь может открыть приветственную пане...   
3  Если пункт **«Приложения»** в ленте недоступен...   
4                              В пункте меню «Файл».   

                                       system_answer  \
0  Функциональное назначение CAD‑модуля — обеспеч...   
1  Минимальные требования для установки и работы ...   
2  Пользователь может открыть приветственную пане...   
3  Диалог управления дополнительными модулями CAD...   
4                              В пункте меню «Файл».   

                                    retrived_context  
0  Программный модуль CAD разработан на базе инте...  
1  Достаточными условиями выполнения программного...  
2  Загрузка и запуск программного модуля CAD осущ...  
3  Данное окно предназначено для осуществления уп...  
4  Перечень инструментов для работы с документами...

In [41]:
llm = ChatOpenAI(
    openai_api_key=MODEL_CONFIG["api_key"],
    openai_api_base=MODEL_CONFIG["base_url"],
    model=MODEL_CONFIG["model_name"],
    temperature=0,
    max_tokens=4000,
    request_timeout=120,
    max_retries=3
)

In [6]:
embeddings = HuggingFaceEmbeddings(
    model_name="intfloat/multilingual-e5-large"
)

/tmp/ipython-input-2105460628.py:1: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embeddings = HuggingFaceEmbeddings(


In [12]:
ragas_data = []

for i in range(len(data)):
    # Подготавливаем контексты как список строк (даже если один контекст)
    contexts = [data.iloc[i]['retrived_context']] if data.iloc[i]['retrived_context'] else []

    ragas_sample = {
        'question': data.iloc[i]['question'],
        'answer': data.iloc[i]['system_answer'],
        'contexts': contexts,
        'ground_truth': data.iloc[i]['golden_answer']
    }
    ragas_data.append(ragas_sample)

# Конвертируем в Dataset HuggingFace
dataset = Dataset.from_list(ragas_data)

In [14]:
metrics = [
    faithfulness,  # Фактическая точность относительно контекста
    answer_relevancy,  # Релевантность ответа вопросу
    context_precision,  # Точность ретривера (насколько релевантны извлеченные контексты)
    context_recall,  # Полнота извлечения (требует ground_truth)
]

In [53]:
run_config = RunConfig(
    max_workers=4,
    timeout=180
)

In [15]:
result = evaluate(
    dataset=dataset,
    metrics=metrics,
    llm=llm,
    embeddings=embeddings,
    run_config=run_config
)

Evaluating:   0%|          | 0/160 [00:00<?, ?it/s]

ERROR:ragas.executor:Exception raised in Job[2]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[38]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[72]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[120]: LLMDidNotFinishException(The LLM generation was not completed. Please increase the max_tokens and try again.)
ERROR:ragas.executor:Exception raised in Job[138]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[148]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[144]: APIConnectionError(Connection error.)


In [23]:
type(result)

ragas.dataset_schema.EvaluationResult

In [26]:
result

{'faithfulness': 0.7710, 'answer_relevancy': 0.9197, 'context_precision': 0.9730, 'context_recall': 0.8415}

In [36]:
metrics = [
    faithfulness,
    answer_relevancy,
    context_precision,
    context_recall,
    answer_correctness,
    answer_similarity
]

In [37]:
result = evaluate(
    dataset=dataset,
    metrics=metrics,
    llm=llm,
    embeddings=embeddings,
    run_config=run_config
)

Evaluating:   0%|          | 0/240 [00:00<?, ?it/s]

ERROR:ragas.executor:Exception raised in Job[52]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[162]: APIConnectionError(Connection error.)
ERROR:ragas.executor:Exception raised in Job[184]: LLMDidNotFinishException(The LLM generation was not completed. Please increase the max_tokens and try again.)


In [38]:
result

{'faithfulness': 0.7511, 'answer_relevancy': 0.9222, 'context_precision': 0.9500, 'context_recall': 0.8396, 'answer_correctness': 0.9021, 'answer_similarity': 0.9751}

In [40]:
type(result)

ragas.dataset_schema.EvaluationResult

In [51]:
result["faithfulness"][:10]

[1.0,
 1.0,
 1.0,
 0.46153846153846156,
 1.0,
 0.8888888888888888,
 0.8,
 0.3333333333333333,
 1.0,
 1.0]

#### BLEU



In [30]:
smoothie = SmoothingFunction().method4  # Сглаживание для коротких текстов

bleu_scores = []
for i in range(len(data)):
    reference = [data.iloc[i]['golden_answer'].split()]  # BLEU требует список списков слов
    candidate = data.iloc[i]['system_answer'].split()

    if reference and candidate:
        score = sentence_bleu(reference, candidate, smoothing_function=smoothie)
        bleu_scores.append(score)
    else:
        bleu_scores.append(0.0)

In [31]:
average_bleu = sum(bleu_scores) / len(bleu_scores)
print(f"Average BLEU Score: {average_bleu:.4f}")

Average BLEU Score: 0.3475


#### ROUGE

In [32]:
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

In [33]:
rouge1_scores = []
rouge2_scores = []
rougeL_scores = []

for i in range(len(data)):
    reference = data.iloc[i]['golden_answer']
    candidate = data.iloc[i]['system_answer']

    if reference and candidate:
        scores = scorer.score(reference, candidate)
        rouge1_scores.append(scores['rouge1'].fmeasure)
        rouge2_scores.append(scores['rouge2'].fmeasure)
        rougeL_scores.append(scores['rougeL'].fmeasure)
    else:
        rouge1_scores.append(0.0)
        rouge2_scores.append(0.0)
        rougeL_scores.append(0.0)

average_rouge1 = sum(rouge1_scores) / len(rouge1_scores)
average_rouge2 = sum(rouge2_scores) / len(rouge2_scores)
average_rougeL = sum(rougeL_scores) / len(rougeL_scores)

print(f"Average ROUGE-1: {average_rouge1:.4f}")
print(f"Average ROUGE-2: {average_rouge2:.4f}")
print(f"Average ROUGE-L: {average_rougeL:.4f}")

Average ROUGE-1: 0.5011
Average ROUGE-2: 0.2823
Average ROUGE-L: 0.5011


Построен синтетический набор данных для использования библиотеки RAGAS (столбцы: question, answer, contexts, ground_truth)

Выполнен расчёт метрик:

- Faithfulness: 0.7511 - Хорошая фактическая точность

- Answer Relevancy: 0.9222 - Высокая релевантность

- Context Precision: 0.9500 - Высокая точность ретривера

- Context Recall: 0.8396 - Хорошая полнота

- Answer Correctness: 0.9021 - Высокая корректность ответов

- Answer Similarity: 0.9751 - Высокое сходство с эталоном